### ISSUE TITLE
Upgrade StorySniffer from 1.0.7 to 1.0.9 and Remove scikit-learn Version Pin

### ISSUE THREAD

**Issue by dkiesow** [LocalNewsImpact/MizzouNewsCrawler#119]
## Overview
StorySniffer has released two new versions (1.0.8 and 1.0.9) since our current version 1.0.7. This upgrade will allow us to remove the `scikit-learn==1.5.1` version pin that was required due to StorySniffer's previous use of pickle files for model serialization.

## Current State
- **Current StorySniffer version:** 1.0.7
- **Latest StorySniffer version:** 1.0.9
- **Current scikit-learn pin:** 1.5.1 (pinned for StorySniffer compatibility)
- **Latest scikit-learn version:** 1.7.2

### Version 1.0.7 (Current)
- **Key change:** Switched from pickle files to `skops.io` for model serialization ([PR #30](https://github.com/palewire/storysniffer/pull/30))
- Updated scikit-learn from 1.4.2 to 1.5.0
- This version introduced the skops library which provides safer, more version-independent model serialization

### Version 1.0.8
- Minor updates and dependency bumps
- Full changelog: https://github.com/palewire/storysniffer/compare/1.0.7...1.0.8

### Version 1.0.9 (Latest - Released 4 minutes ago as of Oct 31, 2025)
- Latest release with most recent updates
- Full changelog: https://github.com/palewire/storysniffer/compare/1.0.8...1.0.9

### The Pickle Problem (Historical Context)
In version 1.0.6 and earlier, StorySniffer used Python pickle files to serialize its ML models. Pickle files are tightly coupled to the exact scikit-learn version used during training, which forced us to pin `scikit-learn==1.5.1` to prevent model deserialization errors.

### The Skops Solution (Current)
Starting in version 1.0.7, StorySniffer switched to `skops.io` for model serialization. Skops is specifically designed for scikit-learn model persistence and provides:
- **Version independence:** Models trained with one scikit-learn version can be loaded with compatible versions
- **Security:** Safer than pickle, with explicit trust declarations
- **Compatibility:** Designed to work across scikit-learn minor version changes

### Benefits of Upgrading
1. **Remove scikit-learn pin:** Can upgrade to latest scikit-learn (1.7.2) or allow flexible versioning
2. **Latest StorySniffer improvements:** Get bug fixes and enhancements from 1.0.8 and 1.0.9
3. **Future-proof:** Skops-based models are more resilient to dependency updates
4. **Security:** Latest dependencies include security patches

### Phase 1: Local Testing
1. **Create test environment:**
   ```bash
   python -m venv test-env
   source test-env/bin/activate
   pip install storysniffer==1.0.9
   ```

2. **Verify StorySniffer functionality:**
   ```python
   from storysniffer import StorySniffer
   sniffer = StorySniffer()
   
   # Test with known article URL
   result = sniffer.guess("https://columbiamissourian.com/news/local/test-article.html")
   print(f"Result: {result}")  # Should return True
   
   # Test with non-article URL
   result = sniffer.guess("https://example.com/about/")
   print(f"Result: {result}")  # Should return False
   ```

3. **Check scikit-learn compatibility:**
   ```bash
   # Try with different scikit-learn versions
   pip install scikit-learn==1.5.1  # Current pinned version
   python test_storysniffer.py
   
   pip install scikit-learn==1.6.0  # Newer version
   python test_storysniffer.py
   
   pip install scikit-learn==1.7.2  # Latest version
   python test_storysniffer.py
   ```

4. **Verify model loading:**
   ```python
   # Ensure models load without warnings/errors
   import warnings
   warnings.simplefilter("error")  # Turn warnings into errors for testing
   
   sniffer = StorySniffer()
   # Should complete without errors
   ```

### Phase 2: Integration Testing
1. **Test with existing verification service:**
   ```bash
   # Run verification on sample URLs
   python -m src.cli.cli_modular verify-urls --batch-size 10 --max-batches 1
   ```

2. **Verify telemetry output:**
   - Check logs for any model loading warnings
   - Verify classification accuracy remains consistent
   - Ensure processing speed is similar or better

3. **Compare classification results:**
   - Run verification on same 100 URLs before and after upgrade
   - Compare results to ensure consistency (should be >95% match)
   - Investigate any significant differences

### Phase 3: Staging Deployment
1. Update requirements files with new versions
2. Rebuild Docker images (base, crawler, processor)
3. Deploy to staging environment
4. Run full pipeline for 24 hours
5. Monitor:
   - Error rates in verification step
   - StorySniffer classification distribution
   - Memory usage (model loading)
   - Processing times

### Phase 4: Production Validation
1. Deploy to production
2. Monitor for 48 hours:
   - Verification success rate (target: >98%)
   - Article vs non-article ratio (expect ~97% articles per historical data)
   - Error logs for any sklearn/storysniffer warnings
3. Compare metrics to pre-upgrade baseline

### Step 1: Update Requirements Files
Update all relevant requirements files:

**requirements-base.txt:**
```txt
storysniffer>=1.0.9  # URL verification service (skops-based models)
scikit-learn>=1.5.1  # Compatible with StorySniffer 1.0.9+
```

**requirements-crawler.txt:**
```txt
scikit-learn>=1.5.1  # Flexible version, compatible with StorySniffer
storysniffer>=1.0.9  # News article classification
```

**requirements-ml.txt:**
```txt
scikit-learn>=1.5.1  # Flexible version for ML operations
storysniffer>=1.0.9  # News article classification
```

**requirements.txt:**
```txt
scikit-learn>=1.5.1  # Flexible version, skops-compatible
storysniffer>=1.0.9  # Compatible version with skops models
```

### Step 2: Update Comments
Replace all comments that mention "Fixed version for StorySniffer compatibility" or "Pinned for StorySniffer pickle" with "Compatible with StorySniffer skops models"

### Step 3: Testing Script
Create `tests/test_storysniffer_upgrade.py`:
```python
"""Test StorySniffer upgrade compatibility."""
import pytest
from storysniffer import StorySniffer

class TestStorySnifferUpgrade:
    @pytest.fixture
    def sniffer(self):
        return StorySniffer()
    
    def test_article_urls(self, sniffer):
        """Test known article URLs are classified correctly."""
        article_urls = [
            "https://columbiamissourian.com/news/local/article-slug/article_abc123.html",
            "https://www.komu.com/news/midmissourinews/story-title-here",
            "https://columbiatribune.com/story/news/local/2024/01/15/article-title",
        ]
        for url in article_urls:
            assert sniffer.guess(url) is True, f"Failed for {url}"
    
    def test_non_article_urls(self, sniffer):
        """Test known non-article URLs are classified correctly."""
        non_article_urls = [
            "https://example.com/about/",
            "https://example.com/contact",
            "https://example.com/category/news/",
            "https://example.com/static/logo.png",
        ]
        for url in non_article_urls:
            assert sniffer.guess(url) is False, f"Failed for {url}"
    
    def test_with_text(self, sniffer):
        """Test classification with text hints."""
        url = "https://example.com/2024/01/15/news-article/"
        text = "Local officials announce new policy"
        assert sniffer.guess(url, text) is True
    
    def test_model_loading(self):
        """Test that models load without errors."""
        import warnings
        with warnings.catch_warnings():
            warnings.simplefilter("error")
            sniffer = StorySniffer()
            assert sniffer.path_only_model is not None
            assert sniffer.path_and_text_model is not None
```

### Step 4: Version Compatibility Matrix
Test the following combinations locally:

| StorySniffer | scikit-learn | Status | Notes |
|--------------|--------------|--------|-------|
| 1.0.7 | 1.5.1 | ✅ Current | Baseline |
| 1.0.9 | 1.5.1 | 🧪 Test | Upgrade StorySniffer only |
| 1.0.9 | 1.6.0 | 🧪 Test | Upgrade both (minor) |
| 1.0.9 | 1.7.2 | 🧪 Test | Upgrade both (latest) |

### Step 5: Docker Image Updates
1. **Base Image:** Update requirements-base.txt → rebuild `mizzou-base`
2. **ML Base Image:** Update requirements-ml.txt → rebuild `mizzou-ml-base`
3. **Crawler Image:** Depends on base → rebuild
4. **Processor Image:** Depends on ml-base → rebuild

Build order:
```bash
# Option A: Conservative (just upgrade StorySniffer)
gcloud builds triggers run build-base-manual --branch=main
gcloud builds triggers run build-ml-base-manual --branch=main
# Wait for completion, then:
gcloud builds triggers run build-crawler-manual --branch=main
gcloud builds triggers run build-processor-manual --branch=main

# Option B: Aggressive (upgrade both to latest)
# Same build order, but with scikit-learn>=1.7.0
```

### Step 6: Rollback Plan
If issues are detected:

1. **Immediate rollback:**
   ```bash
   kubectl set image deployment/mizzou-processor \
     processor=gcr.io/mizzou-news-crawler/mizzou-processor:previous-tag
   kubectl set image deployment/mizzou-crawler \
     crawler=gcr.io/mizzou-news-crawler/mizzou-crawler:previous-tag
   ```

2. **Revert requirements:**
   ```bash
   git revert <commit-sha>
   git push origin main
   ```

3. **Rebuild with old versions:**
   Use previous working image tags or rebuild from reverted commit

### Option A: Conservative (Recommended)
1. Upgrade StorySniffer to 1.0.9
2. Keep scikit-learn at 1.5.1 initially
3. Monitor for 1 week
4. If stable, upgrade scikit-learn to 1.6.x or remove pin

**Pros:**
- Minimal risk
- Easier to isolate issues
- Faster rollback if needed

**Cons:**
- Two-phase upgrade process
- Doesn't immediately remove pin

### Option B: Moderate
1. Upgrade StorySniffer to 1.0.9
2. Change scikit-learn pin to `>=1.5.1,<1.7` (allow 1.6.x)
3. Monitor for 1 week
4. If stable, remove upper bound

**Pros:**
- Gets some scikit-learn updates
- Still constrained to minor versions
- Good middle ground

**Cons:**
- Still has constraints
- May need future adjustment

### Option C: Aggressive
1. Upgrade both StorySniffer to 1.0.9 and scikit-learn to 1.7.2
2. Monitor closely

**Pros:**
- Gets all latest updates immediately
- Maximum benefit from skops compatibility

**Cons:**
- Higher risk of unexpected issues
- Harder to debug if problems occur
- More significant change

## Success Metrics
- ✅ StorySniffer loads without warnings
- ✅ Verification success rate ≥ 98%
- ✅ Article classification rate ~97% (consistent with baseline)
- ✅ No increase in error logs
- ✅ Processing time ≤ baseline (currently ~2.7ms per URL)
- ✅ Memory usage stable or improved

## Timeline (Option A - Conservative)
- **Week 1:** Local testing and requirements updates
- **Week 2:** Staging deployment and monitoring
- **Week 3:** Production deployment of StorySniffer upgrade
- **Week 4:** Monitor and validate
- **Week 5:** Remove scikit-learn pin if stable
- **Week 6:** Final validation

## Related Issues
- #118 - Enable Database Telemetry for URL Verification Service (dependent on stable verification)

## References
- StorySniffer GitHub: https://github.com/palewire/storysniffer
- StorySniffer 1.0.7 Release: https://github.com/palewire/storysniffer/releases/tag/1.0.7
- StorySniffer 1.0.9 Release: https://github.com/palewire/storysniffer/releases/tag/1.0.9
- Skops Documentation: https://skops.readthedocs.io/
- scikit-learn Releases: https://scikit-learn.org/stable/whats_new.html
